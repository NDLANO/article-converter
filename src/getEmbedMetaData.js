/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import defined from 'defined';

export default function getEmbedMetaData(embeds) {
  return embeds.reduce((ctx, embed) => {
    const key = `${embed.data.resource}s`;
    const resourceMetaData = defined(ctx[key], []);
    const embedPlugin = embed.plugin;

    if (
      embedPlugin &&
      embed.status !== 'error' &&
      embedPlugin.getMetaData?.(embed)
    ) {
      const metaData = embedPlugin.getMetaData(embed);
      return {
        ...ctx,
        [key]: [...resourceMetaData, metaData],
      };
    } else {
      return ctx;
    }
  }, {});
}
