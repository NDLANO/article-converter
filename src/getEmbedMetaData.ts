/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import defined from 'defined';
import { EmbedType, LocaleType, PluginUnion } from './interfaces';
import { findPlugin } from './utils/findPlugin';

export default function getEmbedMetaData(
  embeds: EmbedType[],
  locale: LocaleType,
  plugins: PluginUnion[],
) {
  return embeds.reduce((ctx: Record<string, unknown[]>, embed) => {
    const plugin = findPlugin(plugins, embed);
    const key = `${embed.data.resource}s`;
    const resourceMetaData = defined(ctx[key], []);
    const metaData = plugin?.getMetaData?.(embed, locale);
    if (embed.status !== 'error' && metaData) {
      return {
        ...ctx,
        [key]: [...resourceMetaData, metaData],
      };
    } else {
      return ctx;
    }
  }, {});
}
