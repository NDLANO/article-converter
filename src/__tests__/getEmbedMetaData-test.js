/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import getEmbedMetaData from '../getEmbedMetaData';
import createImagePlugin from '../plugins/imagePlugin';
import createAudioPlugin from '../plugins/audioPlugin';
import createBrightcovePlugin from '../plugins/brightcovePlugin';

const imagePlugin = createImagePlugin();
const audioPlugin = createAudioPlugin();
const brightcovePlugin = createBrightcovePlugin();

test('get meta data from embeds', async () => {
  const embeds = [
    {
      data: { resource: 'image' },
      image: {
        title: { title: 'title' },
        alttext: { alttext: 'alt' },
        copyright: { license: 'by-sa' },
        imageUrl: 'http://example.no/1234.jpg',
      },
      plugin: imagePlugin,
    },
    {
      data: { resource: 'image' },
      image: {
        title: { title: 'title 2' },
        alttext: { alttext: 'alt 2' },
        copyright: { license: 'by-sa' },
        imageUrl: 'http://example.no/4123.jpg',
      },
    },
    {
      data: { resource: 'audio' },
      audio: {
        title: { title: 'Tittel' },
        language: 'nb',
        audioFile: {
          url: 'http://audio.no/file/voof.mp3',
        },
        copyright: { license: 'by-sa' },
      },
      plugin: audioPlugin,
    },
    {
      data: { resource: 'brightcove', account: 'account', videoid: '1337' },
      brightcove: {
        name: 'video title',
        copyright: { license: 'by-sa' },
        images: {
          poster: {
            src: 'https://image.no/123.jpg',
          },
        },
        sources: [
          {
            src: 'test3.mp4',
            container: 'STREAM',
            size: 3,
            width: 1000,
            height: 600,
          },
          {
            src: 'test4.mp4',
            container: 'MP4',
            size: 4,
          },
        ],
      },
      plugin: brightcovePlugin,
    },
    {
      data: { resource: 'brightcove', account: 'account', videoid: '42' },
      brightcove: {
        copyright: { license: 'by-sa' },
        sources: [
          {
            src: 'test3.mp4',
            container: 'MP4',
            size: 3,
          },
          {
            src: 'test4.mp4',
            container: 'MP4',
            size: 4,
          },
        ],
      },
      plugin: brightcovePlugin,
    },
  ];

  const copyrights = getEmbedMetaData(embeds);

  expect(copyrights).toMatchSnapshot();
});
