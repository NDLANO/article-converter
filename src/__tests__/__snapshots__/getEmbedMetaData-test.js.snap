// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get meta data from embeds 1`] = `
Object {
  "audios": Array [
    Object {
      "copyright": Object {
        "license": "by-sa",
      },
      "src": "http://audio.no/file/voof.mp3",
      "title": "Tittel",
    },
  ],
  "brightcoves": Array [
    Object {
      "copyright": Object {
        "license": "by-sa",
      },
      "cover": "https://image.no/123.jpg",
      "iframeSrc": "https://players.brightcove.net/account/default_default/index.html?videoId=1337",
      "src": "test4.mp4",
      "title": "video title",
    },
    Object {
      "copyright": Object {
        "license": "by-sa",
      },
      "cover": undefined,
      "iframeSrc": "https://players.brightcove.net/account/default_default/index.html?videoId=42",
      "src": undefined,
      "title": undefined,
    },
  ],
  "images": Array [
    Object {
      "altText": "alt",
      "copyright": Object {
        "license": "by-sa",
      },
      "src": "http://example.no/1234.jpg",
      "title": "title",
    },
  ],
}
`;
